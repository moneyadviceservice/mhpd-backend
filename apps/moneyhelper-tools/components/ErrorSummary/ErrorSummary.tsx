import { H5, Icon, IconType } from '@maps-digital/shared/ui';
import Link from 'next/link';
import {
  forwardRef,
  useEffect,
  useImperativeHandle,
  useState,
  useRef,
} from 'react';
import { twMerge } from 'tailwind-merge';

type Props = {
  title: string;
  errors: Record<string, string[]>;
  classNames?: string;
  errorKeyPrefix?: string;
};
export type Ref = {
  focus: () => void;
  scrollIntoView: () => void;
} | null;

const scrollElementIntoView = (element: HTMLElement | null | undefined) => {
  if (element) {
    element.scrollIntoView({
      behavior: 'smooth',
      block: 'start',
      inline: 'nearest',
    });
  }
};

export const ErrorSummary = forwardRef<Ref, Props>(
  ({ title, errors, classNames, errorKeyPrefix = '' }: Props, ref) => {
    const errorRef = useRef<HTMLDivElement | null>(null);
    useImperativeHandle(
      ref,
      () => {
        return {
          focus() {
            errorRef?.current?.focus();
          },
          scrollIntoView() {
            scrollElementIntoView(errorRef?.current);
          },
        };
      },
      [],
    );
    const [jsEnabled, setJsEnabled] = useState(false);

    const handleScroll = (key: string) => {
      const target = document.getElementById(key);
      const parent = target?.parentElement?.parentElement;

      const y = target?.getBoundingClientRect().top;
      const parentY = parent?.getBoundingClientRect().top;

      if (y && parentY) {
        window.scroll(0, y - (y - parentY));
        target.focus();
      }
    };

    useEffect(() => {
      setJsEnabled(true);
    }, []);

    return (
      <div
        ref={errorRef}
        role="alert"
        aria-relevant="all"
        aria-describedby="error-summary-heading"
        id="error-summary-heading"
      >
        {Object.keys(errors).length > 0 && (
          <div
            className={twMerge(
              't-error-summary border-4 w-full border-red-700 py-8 px-4 sm:px-6 lg:px-8 outline-none',
              classNames,
            )}
          >
            <div className="flex items-center">
              <Icon
                data-testid="icon-error"
                className="min-w-[50px]"
                type={IconType.WARNING}
              ></Icon>
              <H5 id="error-summary-heading">{title}</H5>
            </div>
            <div className="pl-4 md:pl-[50px]">
              <ul className="list-disc mt-3 ml-4">
                {Object.keys(errors).map((key) => {
                  return (
                    errors[key].length > 0 && (
                      <li
                        key={key}
                        className={twMerge(
                          Object.keys(errors).length < 2 && 'list-none -ml-4',
                          'underline text-lg decoration-solid text-red-700 mb-2.5 last:mb-0 break-words',
                        )}
                      >
                        <Link
                          className="whitespace-pre-wrap"
                          href={`#${errorKeyPrefix}${key}`}
                          scroll={false}
                          onClick={(e) => {
                            if (jsEnabled) {
                              e.preventDefault();
                              handleScroll(`${errorKeyPrefix}${key}`);
                            }
                          }}
                        >
                          {errors[key]}
                        </Link>
                      </li>
                    )
                  );
                })}
              </ul>
            </div>
          </div>
        )}
      </div>
    );
  },
);

ErrorSummary.displayName = 'ErrorSummary';
