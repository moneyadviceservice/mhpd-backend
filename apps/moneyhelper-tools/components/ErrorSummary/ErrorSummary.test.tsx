import '@testing-library/jest-dom';
import { fireEvent, getByText, render } from '@testing-library/react';
import { ErrorSummary } from './ErrorSummary';

const errors = {
  test1: ['test1', ' - ', 'test1'],
};

window.HTMLElement.prototype.getBoundingClientRect = () =>
  ({
    top: 20,
  } as DOMRect);

window.HTMLElement.prototype.scrollIntoView = jest.fn();
jest.spyOn(window, 'scroll').mockImplementation(() => {});

describe('ErrorSummary component', () => {
  it('renders correctly', () => {
    const { container } = render(
      <>
        <ErrorSummary title={''} errors={errors} />
        <input id="test1" name="test1" type="text" />
      </>,
    );

    fireEvent(
      getByText(container, 'test1 - test1', { exact: false }),
      new MouseEvent('click', {
        bubbles: true,
        cancelable: true,
      }),
    );

    expect(container).toMatchSnapshot();
  });
});
