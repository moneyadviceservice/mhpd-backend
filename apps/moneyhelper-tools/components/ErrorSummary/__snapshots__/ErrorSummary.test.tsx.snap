// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ErrorSummary component renders correctly 1`] = `
<div>
  <div
    aria-describedby="error-summary-heading"
    aria-relevant="all"
    id="error-summary-heading"
    role="alert"
  >
    <div
      class="t-error-summary border-4 w-full border-red-700 py-8 px-4 sm:px-6 lg:px-8 outline-none"
    >
      <div
        class="flex items-center"
      >
        <svgr-mock
          classname="min-w-[50px]"
          data-testid="icon-error"
        />
        <h5
          class="text-lg md:text-xl text-gray-800 font-bold"
          id="error-summary-heading"
        />
      </div>
      <div
        class="pl-4 md:pl-[50px]"
      >
        <ul
          class="list-disc mt-3 ml-4"
        >
          <li
            class="list-none -ml-4 underline text-lg decoration-solid text-red-700 mb-2.5 last:mb-0 break-words"
          >
            <a
              class="whitespace-pre-wrap"
              href="#test1"
            >
              test1
               - 
              test1
            </a>
          </li>
        </ul>
      </div>
    </div>
  </div>
  <input
    id="test1"
    name="test1"
    type="text"
  />
</div>
`;
